midenc-compile/src/stages/cross_ctx/lift_exports.rs

```diff
@@ -0,0 +1,170 @@
+//! lifting the exports for cross-context calls.
+
+use std::collections::BTreeMap;
+
+use miden_assembly::Spanned;
+use midenc_hir::{
+    diagnostics::Severity,
+    pass::AnalysisManager,
+    types::Abi::{self, Canonical},
+    ComponentBuilder, ComponentExport, FunctionType, InstBuilder, InterfaceFunctionIdent,
+};
+use midenc_session::{DiagnosticsHandler, Session};
+
+use super::flat::{
+    assert_core_wasm_signature_equivalence, flatten_function_type, needs_transformation,
+    FlatteningDirection,
+};
+use crate::{stage::Stage, CompilerResult, LinkerInput};
+
+/// Generates lifting for exports for the cross-context calls according to the Miden ABI.
+///
+/// For each component export ensure a module for each interface that has a lifting
+/// function, i.e. a function that takes parameters according to the cross-context Miden
+/// ABI, converts them to the Wasm CABI, calls the core Wasm module exported function,
+/// converts the results to the cross-context Miden ABI
+///
+/// After this stage all exported functons are expected to be called using the Miden ABI for
+/// cross-context calls, i.e. using the stack and the advice provider for arguments and results.
+pub struct LiftExportsCrossCtxStage;
+
+impl Stage for LiftExportsCrossCtxStage {
+    type Input = LinkerInput;
+    type Output = LinkerInput;
+
+    fn run(
+        &mut self,
+        input: Self::Input,
+        _analyses: &mut AnalysisManager,
+        session: &Session,
+    ) -> CompilerResult<Self::Output> {
+        let component = if let LinkerInput::Hir(component) = input {
+            component
+        } else {
+            return Ok(input);
+        };
+
+        // dbg!(&component.exports());
+        // dbg!(&component.modules().keys());
+
+        let mut component_builder = ComponentBuilder::load(*component, &session.diagnostics);
+
+        let mut lifted_exports: BTreeMap<InterfaceFunctionIdent, ComponentExport> = BTreeMap::new();
+        let exports = component_builder.exports().clone();
+        for (id, export) in exports.into_iter() {
+            if let Canonical = export.function_ty.abi() {
```

Comment:
This doesn't account for situations in which someone constructs IR that exports a function from the component with an ABI that is something other than Wasm-oriented, e.g. `SystemV`. To be clear, this isn't something our tools will produce, but someone else lowering to HIR may not do so through the Wasm frontend, and instead emit HIR components directly with whatever calling conventions they find appropriate.

Basically, I think we only ever synthesize these "lifting" functions when they correspond to a `canon lift` declaration, and only then. This is one of the reasons why synthesizing these functions in the frontend, where that information is present, is preferable.

Reply:
I'm not sure I understand it. `Canonical` here means the Miden ABI. That's confusing. It's another type we have for describing the component imports/exports - https://github.com/0xPolygonMiden/compiler/blob/greenhat/i303-cross-ctx-lower/hir-type/src/lib.rs#L597-L608, and it'll be deleted once we migrate to the HIR2. We'll generate the synthetic functions in the frontend only for the `canon lift/lower` ones.

Reply:
Yeah I'm saying we do not need to represent the notion of a "Miden ABI" at all, it is implied that lowering a function to Miden Assembly _by definition_ must adhere to the "Miden ABI". From a call site perspective, it is sufficient to know only the original callee signature (including its calling convention), because the low-level details of how to invoke the resulting Miden Assembly procedure can be completely derived from that signature.

Backing up a bit, and to help clarify conversations about this going forward, here's how I define the distinction between ABIs and calling conventions:

* An _ABI_ describes how a given set of types are laid out in memory, and how procedure calls _in general_ are executed by the machine, such that two applications can even begin to talk to one another. It is a very low-level, machine-dependent specification. For our purposes, ABI consists of:
  - The specification for how byte-addressable memory is laid out in Miden's word-addressable memory, and on the Miden operand stack
  - The bitwise representation of types in the IR type system, i.e. how they are laid out in byte-addressable memory
  - The specification of what constitutes valid _immediates_ in IR type system terms
  - The maximum addressable depth of the operand stack, and how overflow is handled by the VM
  - Operand stack effects of the various procedure invocation instructions: i.e. `exec` has none, `call` implicitly hides/restores the overflow portion of the operand stack and switches context, `dynexec` consumes the first element on the stack as a procedure reference, etc.
  - Program lifecycle, e.g. rodata initialization, global constructors, etc.
* A _calling convention_ encodes the information necessary to understand how to call a given function, i.e. how it expects to receive its arguments, and how you can expect to receive its results. A calling convention is generally specified in terms of a specific ABI. For example, the C calling convention can mean different things depending on the target ABI/architecture. In general, this includes:
  - The set of types specified by that convention
  - The way in which specific types are expected to be passed at the machine level (i.e. is a struct passed by value, or by reference, is there a threshold at which that happens). Is there a distinction between integers and floats, etc.
  - How to handle extension of integer values whose type is smaller than the machine word (i.e. are they sign-extended, or zero-extended). This isn't as important in a stack machine, but in a register machine it is essential, since different machine instructions can read the same register as different bit-widths, and failing to sign/zero-extend a value can result in silent corruption of computed results.
  - The range of possible function effects (e.g. are calls asynchronous, will they return, can they throw exceptions) and how those are handled
* In addition to those more general properties, any calling convention we support must also specify:
  - The procedure call protocol for passing arguments via the advice provider when the call involves a context switch, or when the callee arguments would overflow the addressable stack.
  - Whether the call involves a context switch (can also be considered a function effect)

So, with that in mind, what I'm trying to say is that the notion of a "Miden ABI" is implicit - we do not have any other target than the Miden VM, so by definition all of our calling conventions are/must be specified in terms of how they are represented at the Miden VM level.

So consider the `Wasm` calling convention, it specifies what types are allowed in function signatures using that convention (i.e. any type described in the core Wasm spec). _We_ have further defined how that convention is to be translated into Miden Assembly. When the IR is constructed, we will validate that there are no convention violations (i.e. that there aren't argument types with no corresponding Wasm representation). Only when we move to the code generation stage will we actually need to deal with calling convention details such as spilling of excess arguments, layout of types on the operand stack/in memory, etc.

Consider another example: the `CanonLift` convention. This one is maybe more interesting, because not only does the convention imply a context switch, it also permits a much broader range of types. Because of the context switch, we are required to handle certain types (i.e. heap-allocated types) a certain way, by passing them via the advice provider. We must also account for spilling of excess arguments, but again, due to the context switch we are required to handle spills a specific way as well (via the advice provider). Beyond those details, we have not yet defined specifically how the various allowed types permitted by the convention will be passed at the VM level, but that is entirely our perogative. For the most part, I suspect that we will treat most types the same we do in `Wasm`, as that already accounts for most (all?) scalar types allowed.

Reply:
Whoa! That's gold! I'm certainly guilty of conflating calling convention and ABI many times.

Reply:
Yeah me too, I've been playing fast and loose with the terminology in a lot of these conversations and now we're paying the price 😅